src/callbacks_stage/positivity_shallow_water_dg1d.jl

```diff
@@ -84,4 +86,89 @@ function limiter_shallow_water!(u, threshold::Real, variable,
 
     return nothing
 end
+
+# !!! warning "Experimental code"
+#     This is an experimental feature and may change in future releases.
+
+# Note that for the `ShallowWaterMultiLayerEquations1D` only the waterheight `h` is limited in
+# each layer. Furthermore, a velocity desingularization is applied after the limiting to avoid
+# numerical problems near dry states.
+function limiter_shallow_water!(u, threshold::Real, variable,
+                                mesh::Trixi.AbstractMesh{1},
+                                equations::ShallowWaterMultiLayerEquations1D,
+                                dg::DGSEM, cache)
+    @unpack weights = dg.basis
+
+    Trixi.@threaded for element in eachelement(dg, cache)
+        # Limit layerwise
+        for m in eachlayer(equations)
+            # determine minimum value
+            value_min = typemax(eltype(u))
+            for i in eachnode(dg)
+                u_node = get_node_vars(u, equations, dg, i, element)
+                value_min = min(value_min, variable(u_node, equations)[m])
+            end
+
+            # detect if limiting is necessary
+            value_min < threshold - eps() || continue
```

Comment:
Using this function `eps()` means that this will work in single precision as well, right?

Reply:
I didn't test it, but I would expect that it also works in single precision

Reply:
I expect it should because it is how we set the default tolerances in the equation structs for the wetting and drying.